packages/vm/src/state/transientStorage.ts

```diff
@@ -0,0 +1,85 @@
+import { Address } from 'ethereumjs-util'
+
+export interface TransientStorageOps {
+  storage?: Map<Address, Map<string, Buffer>>
+  checkpoints?: any[]
+}
+
+const copyStorage = (
+  input: Map<Address, Map<string, Buffer>>
+): Map<Address, Map<string, Buffer>> => {
+  const map = new Map()
+  for (const [addr, storage] of input.entries()) {
+    const copy = new Map()
+    for (const [key, value] of storage.entries()) {
+      copy.set(key, Buffer.from(value))
+    }
+    map.set(addr, copy)
+  }
+  return map
+}
+
+export default class TransientStorage {
+  _storage: Map<Address, Map<string, Buffer>>
+  _checkpoints: any[]
+
+  constructor(opts: TransientStorageOps = {}) {
+    this._storage = opts.storage ?? new Map()
+    this._checkpoints = opts.checkpoints ?? []
+  }
+
+  get(addr: Address, key: Buffer): Buffer {
+    const map = this._storage.get(addr)
+    if (!map) {
+      return Buffer.alloc(32, 0x00)
+    }
+    const value = map.get(key.toString('hex'))
+    if (!value) {
+      return Buffer.alloc(32, 0x00)
+    }
+    return value
+  }
+
+  put(addr: Address, key: Buffer, value: Buffer) {
+    if (!this._storage.has(addr)) {
+      this._storage.set(addr, new Map())
+    }
+    const storage = this._storage.get(addr)
+    storage!.set(key.toString('hex'), value)
+  }
+
+  revert() {
+    if (this._checkpoints.length) {
+      const storage = this._checkpoints.pop()
+      this._storage = copyStorage(storage)
+    }
+  }
+
+  commit() {
+    const checkpoint = this._checkpoints.pop()
+    for (const [addr, storage] of checkpoint!.entries()) {
+      const map = this._storage.get(addr)
+      if (map) {
+        for (const [key, value] of storage.entries()) {
+          map.set(key, value)
+        }
+      }
+    }
+  }
+
+  checkpoint(): void {
+    const copy = copyStorage(this._storage)
```

Comment:
This is OK for now, but for production we probably want to update this. The logic would be to keep a list of `Map` items where each time if we `checkpoint` we add an empty map to the list. When you want to get a value at `index` (so `TLOAD`) then traverse the list from the back: for each map attempt to get the key. If no key is returned then the value is thus `0`. On `TSTORE` put the key in the last item of the map. On reverting, pop the last list item. On `commit`, copy all items from the last map to the previous map and pop that map from the list.

If we copy the map each time then I think it can blow up, first fill the map with X items and then call own contract until all gas is consumed. Then if we call to depth `L` we have `X*L` 32-byte items in total in memory. This might be a DoS vector which makes the implementation very slow.

Reply:
@ryanio do you think this approach would be a "correct" one to optimize this?

Reply:
that sounds good to me, agree that copyStorage on every checkpoint would have a detrimental impact on performance

Reply:
I've updated the algorithm to be more similar to how geth does it, let me know what you think. It feels a bit more simple to me doing it this way